src/shared/sam/cli/samCliLocator.ts

```diff
@@ -0,0 +1,156 @@
+/*!
+ * Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+'use strict'
+
+import * as path from 'path'
+import { EnvironmentVariables } from '../../environmentVariables'
+import * as filesystemUtilities from '../../filesystemUtilities'
+
+export interface SamCliLocationProvider {
+    getLocation(): Promise<string | undefined>
+}
+
+export class DefaultSamCliLocationProvider implements SamCliLocationProvider {
+
+    private static SAM_CLI_LOCATOR: BaseSamCliLocator | undefined
+
+    public async getLocation(): Promise<string | undefined> {
+        return DefaultSamCliLocationProvider.getSamCliLocator().getLocation()
+    }
+
+    public static getSamCliLocator(): SamCliLocationProvider {
+        if (!DefaultSamCliLocationProvider.SAM_CLI_LOCATOR) {
+            if (process.platform === 'win32') {
+                DefaultSamCliLocationProvider.SAM_CLI_LOCATOR = new WindowsSamCliLocator()
+            } else {
+                DefaultSamCliLocationProvider.SAM_CLI_LOCATOR = new UnixSamCliLocator()
+            }
+        }
+
+        return DefaultSamCliLocationProvider.SAM_CLI_LOCATOR
+    }
+
+}
+
+abstract class BaseSamCliLocator {
+
+    public constructor() {
+        this.verifyOs()
+    }
+
+    public async getLocation(): Promise<string | undefined> {
+        let location: string | undefined = await this.findFileInFolders(
+            this.getExecutableFilenames(),
+            this.getExecutableFolders()
+        )
+
+        if (!location) {
+            location = await this.getSystemPathLocation()
+        }
+
+        return location
+    }
+
+    protected abstract verifyOs(): void
+    protected abstract getExecutableFilenames(): string[]
+    protected abstract getExecutableFolders(): string[]
+
+    protected async findFileInFolders(
+        files: string[],
+        folders: string[]
+    ): Promise<string | undefined> {
+        const fullPaths: string[] = files.map(
+            file => folders.map(folder => path.join(folder, file))
```

Comment:
Be sure to filter out empty folders, as in the PATH `"/foo/bar/bat;;/foo/bar/baz;"`, either here or at the call site. Otherwise we could end up scanning the user's root directory despite it not being in the PATH.